include/drivers/reset.h

```diff
@@ -0,0 +1,340 @@
+ * @param idx logical index into "resets"
+ * @return static initializer for a struct reset_dt_spec for the property
+ */
+#define RESET_DT_SPEC_GET_BY_IDX(node_id, idx)					\
+	{									\
+		.dev = DEVICE_DT_GET(DT_RESET_CTLR_BY_IDX(node_id, idx)),	\
+		.id = DT_RESET_ID_BY_IDX(node_id, idx)				\
+	}
+
+/**
+ * @brief Equivalent to RESET_DT_SPEC_GET_BY_IDX(node_id, 0).
+ *
+ * @param node_id devicetree node identifier
+ * @return static initializer for a struct reset_dt_spec for the property
+ * @see RESET_DT_SPEC_GET_BY_IDX()
+ */
+#define RESET_DT_SPEC_GET(node_id) \
+	RESET_DT_SPEC_GET_BY_IDX(node_id, 0)
+
+/**
+ * @brief Static initializer for a @p reset_dt_spec from a DT_DRV_COMPAT
+ * instance's Reset Controller property at an index.
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @param idx logical index into "resets"
+ * @return static initializer for a struct reset_dt_spec for the property
+ * @see RESET_DT_SPEC_GET_BY_IDX()
+ */
+#define RESET_DT_SPEC_INST_GET_BY_IDX(inst, idx) \
+	RESET_DT_SPEC_GET_BY_IDX(DT_DRV_INST(inst), idx)
+
+/**
+ * @brief Equivalent to RESET_DT_SPEC_INST_GET_BY_IDX(inst, 0).
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @return static initializer for a struct reset_dt_spec for the property
+ * @see RESET_DT_SPEC_INST_GET_BY_IDX()
+ */
+#define RESET_DT_SPEC_INST_GET(inst) \
+	RESET_DT_SPEC_INST_GET_BY_IDX(inst, 0)
+
+/** @cond INTERNAL_HIDDEN */
+
+/**
+ * API template to get the reset status of the device.
+ *
+ * @see reset_status
+ */
+typedef int (*reset_api_status)(const struct device *dev, uint32_t id, uint8_t *status);
+
+/**
+ * API template to put the device in reset state.
+ *
+ * @see reset_assert
+ */
+typedef int (*reset_api_assert)(const struct device *dev, uint32_t id);
+
+/**
+ * API template to take out the device from reset state.
+ *
+ * @see reset_deassert
+ */
+typedef int (*reset_api_deassert)(const struct device *dev, uint32_t id);
+
+/**
+ * API template to reset the device.
+ *
+ * @see reset_toggle
+ */
+typedef int (*reset_api_toggle)(const struct device *dev, uint32_t id);
+
+/**
+ * @brief Reset Controller driver API
+ */
+__subsystem struct reset_driver_api {
+	reset_api_status status;
+	reset_api_assert assert;
+	reset_api_deassert deassert;
+	reset_api_toggle toggle;
+};
+
+/** @endcond */
+
+/**
+ * @brief Get the reset status
+ *
+ * This function returns the reset status of the device.
+ *
+ * @param dev Reset controller device.
+ * @param id Reset line.
+ * @param status Where to write the reset status.
+ *
+ * @retval 0 On success.
+ * @retval -ENOSYS If the functionality is not implemented by the driver.
+ * @retval -errno Other negative errno in case of failure.
+ */
+__syscall int reset_status(const struct device *dev, uint32_t id, uint8_t *status);
+
+static inline int z_impl_reset_status(const struct device *dev, uint32_t id, uint8_t *status)
+{
+	const struct reset_driver_api *api = dev->api;
+
+	if (!api) {
+		return -ENOSYS;
+	}
```

Comment:
nit: this can be removed, it's a programming error to implement a reset driver without API.

Reply:
Ok, I will remove it.